top/doc/manpages/man/man1/vroute.1
.TH VROUTE 1 "Cplant Myrinet Utilities" "Cplant" \" -*- nroff -*-
.SH NAME
.B vroute 
\- vroute client (for verifying myrinet routes on Cplant)
.SH SYNOPSIS
.B vroute
[\-z no_sleep_secs] [\-b] [\-v] [\-s sendnid] [-r recvnid] [-p portnum] \-from sendhost \-to recvhost

.SH DESCRIPTION
.B vroute
is the client side of the myrinet route 
verification utility on Cplant. It works in conjunction with the
.B vrouted 
server via sockets and TCP/IP to provide detection of one-way 
or two-way pings between myrinet nodes. 
.B vroute 
relies on the Cplant rtscts myrinet driver/module running on the
server side.
.P
.B vroute 
normally identifies the service port on the vrouted server by doing a 
lookup in /etc/services on the local node.
.P
An invocation of 
.B vroute 
runs on three nodes: the vroute
client runs on a controller node (SSS-0 or SSS-1) and coordinates 
myrinet pings
between a pair of sender and receiver targets both connected
by myrinet and running 
the
.B vrouted
service.
.P
Unless the -p option is used, all three nodes need to have a line like
.P
vroute        8011/tcp
.P
in their /etc/services file. Here, "8011" binds a specific
port to vrouted service requests. The actual port number
should be chosen as appropriate for a given site. In addition,
the 
.B vroute
client should have entries for the "sendhost" and the "recvhost"
in its /etc/hosts file unless they are registered with a DNS service.
.P
As an alternative to looking up the port number in /etc/services
.B vroute 
can be given an alternate port number using
the -p option. This option can be used if one knows the port number
of the registered service or if one wishes to contact user-invocations
of 
.B vrouted 
on the sender and receiver. In the latter case, the
.B vrouted
servers should be run by hand on the sender and receiver
using the same option (and port number).
.P
.B vroute
attempts to open a socket connection to the sender and
receiver at the 
.B vrouted
service port. If successful, it tells these nodes which
one is to act as sender and which one is to act as receiver
for a one-way myrinet ping (a reverse direction ping may
follow). Optionally, it attempts to verify the user's notion
of the sender's and receiver's Cplant node id (nid) by querying
the sender or receiver for their nids and comparing these with
"sendnid" and "recvnid".
.P
In any case, 
.B vroute 
obtains the sender's nid, giving it to the receiver and
visa versa. (These are the servers' nids obtained from
their portals device, not the input nids to vroute).
.P
.B vroute
then tells the sender to initiate
a myrinet ping and subsequently asks the receiver to check
on receipt of the ping. The sender will try to send a short
myrinet packet using the route in the slot in its MCP's
route table indexed by the receiver nid it just received.
The myrinet message also has the sender's node id and a 
"ping" message type. The receipt of such a message is
handled by the rtscts module -- its arrival is noted by
marking a slot, indexed by the send id, in the receiver's ping_info table.
When the receiver is notified by the client to look for receipt
of a message from the sender it checks this slot in the ping_info table,
and reports success or failure to the vroute client.
.P
When contacted by 
.B vroute 
the servers attempt to open
a pair of devices, /dev/portals and /dev/rtscts, on the sender
and receiver nodes. Opening these devices requires that they exist as files
with the appropriate permissions (normally read and write by
world) and that the corresponding Linux kernel modules be loaded
on the servers.
It is
also assumed that the rtscts MCP (Myrinet Control Program) is 
running on the local lanai card of each one of the servers.
.P
The 
.B vroute 
client does not need to be running portals or rtscts/myrinet.
It just needs to be able to make a TCP/IP connection to the
sender and receiver.

.SS OPTIONS
.TP
.I "\-b"
Both. Attempt a ping in "both" directions; from sendhost to
recvhost and then from recvhost to sendhost. Ignored if sendhost = recvhost.
.TP
.I "\-v"
Verbosity. Print out progress reports for the vroute protocol.
.TP
.I "\-p portnum"
Specify a port number for connecting to the servers. For avoiding
reference to /etc/services if the vrouted service port number is
known or if one wishes to specify an alternative port number in
conjunction with use of the same option in a manual invocation of 
the servers.
When this option is NOT used, 
.B vroute 
will try to
look up the port number for vroute service in its local copy of /etc/services.
.TP
.I "\-s sendnid"
Ask the sender for verification of the proposed sendnid (send node id).
.TP
.I "\-r recvnid"
Ask the receiver for verification of the proposed recvnid (receive node id).
Ignored if sendhost = recvhost.
.TP
.I "\-from sendhost"
Required. Open a socket connection to the sendhost and ask it to send a
myrinet ping to a recvhost. Can be a hostname or an IP address.
If sendhost = recvhost then only one server connection is made and -r and
-b options are ignored.
.TP
.I "\-to recvhost"
Required. Open a socket connection to the recvhost and ask it to check for
receipt of a myrinet ping from the sendhost. Can be a hostname or an IP address.
If sendhost = recvhost then only one server connection is made and -r and
-b options are ignored.
.TP
.I "\-z no_sleep_secs"
Number of seconds to sleep between time that we ask sender to ping and
receiver to look for its arrival. The default is 1 (should be enough).

.SH SUGGESTED USAGE
Call 
.B vroute
from a perl script that does your site's nid-to-hostname mapping,
checking vroute's exit code for success or failure.
.P
A good sanity check in case a ping from node A to node B fails is
to try pings from A to A and from B to B to make sure
the Lanai cards are working, the MCP is running, and the host
is connected to a Myrinet switch. It's pretty hard to screw up
the route from and to the same host (the route is 0x80 followed by all
zeros...).
.P
Using a shell script also makes it easy to run the vroute client and
servers "by hand" in conjunction with use of the -p option (for 
specifying a service port). 
.P
Use of the verbosity option (-v) is sometimes useful in diagnosing the 
case that the vroute protocol fails.
.P
A sample Perl script follows.
.br
.nf
#! /usr/bin/perl

  # su.pl -- for each node in the given su, verify the
  #          routes to all the nodes in the range of
  #          sus specified below


  #-----------------------------------------------------

  # set these site specific parameters

  $MINSU = 13;
  $MAXSU = 21;
  $SUSIZE = 16;

  #-----------------------------------------------------

  $MIN_NODE = $MINSU * $SUSIZE;
  $MAX_NODE = $MAXSU * $SUSIZE + $SUSIZE - 1;
 
  $su_in = $ARGV[0];

  $MIN_SUNODE = $su_in * $SUSIZE;
  $MAX_SUNODE = $MIN_SUNODE + $SUSIZE - 1;

  # verify routes from each node in given su to
  # nodes in SUs hardcoded above

for ($k=$MIN_SUNODE; $k<=$MAX_SUNODE; $k++) {

  $vnode = nmap($k);

  $j = 0;
  $su = $MINSU;

  system("echo '-------------------------------------------------' >> ./bad_nodes_su$su_in");
  system("echo 'testing routes of $vnode...' >> ./bad_nodes_su$su_in");

  for ( $i=$MIN_NODE; $i<=$MAX_NODE; $i++ ) {

    if ( $su >= $MINSU && $su <= $MAXSU ) {
      $name = nmap($i);
      print("vroute -from $vnode -to $name\n");
      $ec = system("/usr/local1/jsotto/vroute -from $vnode -to $name");

      if ( $ec ne 0 ) {
        $msg = "test FAILED: $vnode to $name ($i)";
        print("$msg\n");
        system("echo '$msg' >> ./bad_nodes");
      }
    }

    $j++;
    if ($j == $SUSIZE ) {
      $j = 0;
      $su++;
    }

  }
}

# map physical node id to lexico node name
sub nmap
{
  $id = $_[0];
  $sunum = int($id/$SUSIZE);
  $nid   =  $id % $SUSIZE;
  return "c-$nid.SU-$sunum";
}
.fi

.SH EXIT CODE
Zero if all the proposed tests succeed.
Nonzero otherwise.
So if we do
.P
% vroute -s 11 -from c-5.SU-1 -to c-9.SU-2 -b 
.P
and both the pings succeed, but the sendnid verification fails
then 
.B vroute
will return a nonzero value. On the other hand, vroute's textual
output will report success of the other tests.

.SH VERBOSITY
.B vroute
is always fairly verbose; it reports on success or failure of various
stages of the proposed tests. Verbosity can be bumped up a little with
the -v command line option.

.SH FILES
/etc/services, /etc/hosts

.SH SEE ALSO
vrouted(8), inetd(8)


top/compute/OS/Myrinet/vroute/vroute.c
/*************************************************************************
Cplant Release Version 2.0.1.10
Release Date: Nov 5, 2002 
#############################################################################
#
#     This Cplant(TM) source code is the property of Sandia National
#     Laboratories.
#
#     This Cplant(TM) source code is copyrighted by Sandia National
#     Laboratories.
#
#     The redistribution of this Cplant(TM) source code is subject to the
#     terms of the GNU Lesser General Public License
#     (see cit/LGPL or http://www.gnu.org/licenses/lgpl.html)
#
#     Cplant(TM) Copyright 1998, 1999, 2000, 2001, 2002 Sandia Corporation. 
#     Under the terms of Contract DE-AC04-94AL85000, there is a non-exclusive
#     license for use of this work by or on behalf of the US Government.
#     Export of this program may require a license from the United States
#     Government.
#
#############################################################################
**************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <portals/sys_portal_lib.h>

#include <sys/file.h>
#include <sys/ioctl.h>

#include <getopt.h>

#include "vroute.h"

extern char* optarg;

static struct option opts[]=
{
  {"help",          no_argument,           0,  'h'}, /* help */
  {"snid",          required_argument,     0,  's'}, /* from nid */
  {"rnid",          required_argument,     0,  'r'}, /* to   nid */
  {"from",          required_argument,     0,  'f'}, /* from hostname */
  {"to",            required_argument,     0,  't'}, /* to   hostname */
  {"port",          required_argument,     0,  'p'}, /* maybe vrouted's port */
  {"verbosity",     no_argument,           0,  'v'}, /* verbosity */
  {"both",          no_argument,           0,  'b'}, /* ping both directions */
  {"z",             required_argument,     0,  'z'}, /* sleep this many seconds */
  {0, 0, 0, 0} 
};

int sock_init(char* hostname, int try_port, int verbosity);
void open_dev(int sd0, int sd1, char buf[], int bufsize, char* str);
int nid_check(int sd0, char buf[], int bufsize, int my_nid, 
               int* nid, char* str, int verify);
int do_ping(int send_sd, int recv_sd, char send_buf[], char recv_buf[], 
            int bufsize, int snid, int rnid, 
            char* send_str, char* recv_str);
int self_ping(int send_sd, char send_buf[], int bufsize, int snid,
              char* send_str);
int check_args(int got_sname, int got_rname);
void show_usage(void);

static int no_sleep_secs=1; /* default value for time we sleep between
                               telling sender to ping and receiver to
                               check for ping */

#define BUFSIZE 100

int main(int argc, char* argv[]) 
{
  int send_sd, recv_sd, ch=0, index;
  int try_port=0;
  int my_snid=0, my_rnid=0, snid, rnid, rc_snid, rc_rnid;
  int rlen, slen;
  int ping_receiver=0, ping_sender=0;
                            
  char recv_buf[BUFSIZE], send_buf[BUFSIZE];

  char* send_host=NULL;
  char* recv_host=NULL;

  int got_snid=0, got_rnid=0, got_sname=0, got_rname=0;

  int both_dirs=0, verbosity=0;

  while ( ch != EOF ) {
    ch = getopt_long_only(argc, argv, "", opts, &index);

    switch(ch) {
      case 'z':
        no_sleep_secs = strtol(optarg, (char**)NULL, 10);
        if (no_sleep_secs < 1) no_sleep_secs = 1; /* the default */
        break;

      case 's':
        my_snid = strtol(optarg, (char**)NULL, 10);
        got_snid = 1;
        break;

      case 'r':
        my_rnid = strtol(optarg, (char**)NULL, 10);
        got_rnid = 1;
        break;

      case 'f':
        send_host = optarg;
        got_sname = 1;
        break;

      case 't':
        recv_host = optarg;
        got_rname = 1;
        break;

      case 'p':
        try_port = strtol(optarg, (char**)NULL, 10);
        break;

      case 'b':
        both_dirs = 1;
        break;

      case 'v':
        printf("setting verbosity");
        verbosity = 1;
        break;

      case 'h':
        show_usage();
        exit(0);
        break;

      case ':':
        printf("option w/o arg\n");
        break;

      default:
    }
  }

  if ( check_args(got_sname, got_rname) < 0 ) {
    return(-1);
  }

  /* if we didn't get a nid on the command line, assign it an
     arbitrary value
  */
  if ( !got_snid ) {
    my_snid = 0;
  }

  if ( !got_rnid ) {
    my_rnid = 0;
  }


  /* if send_host == recv_host then do things a little
     differently...
  */

/*--------------------------------------------------------------------*/

  /* special case: sender pings sender... */

  if ( !strcmp(send_host, recv_host) ) {

    printf("\n------------------------------------------\n");
    printf("vroute: %s pings self...\n\n", send_host);

    if ( got_rnid ) {
      printf("vroute: ignoring -r option\n\n");
    }

    if ( both_dirs ) {
      printf("vroute: ignoring -b option\n\n");
    }

    send_sd = sock_init(send_host,try_port,verbosity); 

    if ( send_sd < 0 ) {
      printf("------------------------------------------\n");
      return(-1);
    }

    if (verbosity) {
      printf("\nvroute: trying initial handshake w/ sender...\n");
      printf("vroute: if this hangs, we may not have vrouted registered w/\n");
      printf("vroute: inetd on the target node...\n");
    }
    write( send_sd, "y", 1 );

    slen = doRead(send_sd, send_buf, BUFSIZE, strlen(handshake_msg)); /* ack */

    if (verbosity) {
      send_buf[strlen(handshake_msg)] = 0;
      printf("vroute: handshake -- %s\n", send_buf);
    }

    if ( strncmp(handshake_msg, send_buf, strlen(handshake_msg)) ) {
      printf("vroute: -- got a hunch we're not talking to vrouted\n");
      printf("vroute: -- could be service is not registered with inetd\n");
      printf("vroute: -- on sender %s...\n", send_host);
      printf("------------------------------------------\n");
      return(-1);
    }

    /* tell the sender that he is sender AND receiver */

    write( send_sd, "b", 1 );
    slen = read(send_sd, send_buf, BUFSIZE);       /* ack */

    if (verbosity) {
      printf("vroute: trying open_dev(%s)...\n", send_host);
    }
    open_dev(send_sd, 0, send_buf, BUFSIZE, send_host);

    rc_snid = nid_check(send_sd, send_buf, BUFSIZE, my_snid, &snid,
                        send_host, got_snid);

    write( send_sd, "y", 1 );            /* proceed */
    slen = read(send_sd, send_buf, BUFSIZE);

    printf("\n");
    ping_receiver = self_ping(send_sd, send_buf, BUFSIZE, snid, send_host);

    printf("------------------------------------------\n");

    if ( ping_receiver == 0 || rc_snid == 0 ) {
      return -1;
    }
    return 0;
  }

/*--------------------------------------------------------------------*/

  
  /* vroute client code -- make socket connection to
     vroute service on a sender AND on a receiver.
     instead of having them talk directly, client coords
     their interactions.
  */


  printf("\n------------------------------------------\n");

  if ( both_dirs ) {
    printf("vroute: %s and %s ping each other (two-way test)\n\n", send_host, recv_host);
  }
  else {
    printf("vroute: %s pings %s (one-way test)\n\n", send_host, recv_host);
  }

/*------------------------------------------------------------------*/
  send_sd = sock_init(send_host,try_port,verbosity); 

  if ( send_sd < 0 ) {
    printf("------------------------------------------\n");
    return(-1);
  }

  recv_sd = sock_init(recv_host,try_port,verbosity); 

  if ( recv_sd < 0 ) {
    write( send_sd, "n", 1 );
    printf("------------------------------------------\n");
    return(-1);
  }

  if (verbosity) {
    printf("\nvroute: trying initial handshake w/ sender and receiver...\n");
    printf("vroute: if this hangs, we may not have vrouted registered w/\n");
    printf("vroute: inetd on one of the target nodes...\n");
  }
  write( send_sd, "y", 1 );
  write( recv_sd, "y", 1 );

  slen = doRead(send_sd, send_buf, BUFSIZE, strlen(handshake_msg)); /* ack */

  if (verbosity) {
    printf("vroute: got handshake from sender...do receiver\n");
  }

  rlen = doRead(recv_sd, recv_buf, BUFSIZE, strlen(handshake_msg)); /* ack */

  if (verbosity) {
    printf("vroute: got handshake from receiver...now verify...\n");
  }

  if (verbosity) {
    send_buf[strlen(handshake_msg)] = 0;
    printf("vroute: handshake -- %s\n", send_buf);
  }

  if ( strncmp(handshake_msg, send_buf, strlen(handshake_msg)) ) {
    printf("vroute: -- got a hunch we're not talking to vrouted\n");
    printf("vroute: -- could be service is not registered with inetd\n");
    printf("vroute: -- on sender %s...\n", send_host);
    write( recv_sd, "n", 1 );            /* kill receiver */
    printf("------------------------------------------\n");
    return(-1);
  }

  if ( strncmp(handshake_msg, recv_buf, strlen(handshake_msg)) ) {
    printf("vroute: -- got a hunch we're not talking to vrouted\n");
    printf("vroute: -- could be service is not registered with inetd\n");
    printf("vroute: -- on receiver %s...\n", recv_host);
    write( send_sd, "n", 1 );            /* kill sender */
    printf("------------------------------------------\n");
    return(-1);
  }

  /* get things started */

  /* tell sender that he is the sender */

  write( send_sd, "s", 1 );
  slen = read(send_sd, send_buf, BUFSIZE);       /* ack */

  /* tell receiver that he is the receiver */

  write( recv_sd, "r", 1 );
  rlen = read(recv_sd, recv_buf, BUFSIZE);       /* ack */

  /* sender opens devices */

  if (verbosity) {
    printf("vroute: trying open_dev(%s)...\n", send_host);
  }
  open_dev(send_sd, recv_sd, send_buf, BUFSIZE, send_host);

  /* check sender's nid */

  rc_snid = nid_check(send_sd, send_buf, BUFSIZE, my_snid, &snid, send_host,
            got_snid);

  /* receiver opens devices */
  printf("\n");

  if (verbosity) {
    printf("vroute: trying open_dev(%s)...\n", recv_host);
  }
  open_dev(recv_sd, send_sd, recv_buf, BUFSIZE, recv_host);

  /* check receiver's nid */

  rc_rnid = nid_check(recv_sd, recv_buf, BUFSIZE, my_rnid, &rnid, recv_host,
            got_rnid);

  write( recv_sd, "y", 1 );            /* proceed */
  write( send_sd, "y", 1 );            /* proceed */
  slen = read(send_sd, send_buf, BUFSIZE);
  rlen = read(recv_sd, recv_buf, BUFSIZE);

  /* sender and receiver should be synchronized */

  /* tell sender receiver's nid */

  itoa(rnid,istr);
  write( send_sd, istr, sizeof(istr) );

  /* tell receiver sender's nid */
  itoa(snid,istr);
  write( recv_sd, istr, sizeof(istr) );

  slen = read(send_sd, send_buf, BUFSIZE);
  rlen = read(recv_sd, recv_buf, BUFSIZE);

  printf("\n");
  ping_receiver = do_ping(send_sd, recv_sd, send_buf, recv_buf, BUFSIZE, 
                          snid, rnid, send_host, recv_host);

  /* try other direction? */

  if ( both_dirs == 1 ) {

    printf("\n");
    ping_sender = do_ping(recv_sd, send_sd, recv_buf, send_buf, BUFSIZE, 
                          rnid, snid, recv_host, send_host);
  }
  else {
    write( recv_sd, "n", 1 );            /* kill receiver */
    write( send_sd, "n", 1 );            /* kill sender */
  }

  printf("------------------------------------------\n");

/*------------------------------------------------------------------*/

  if ( both_dirs ) {
    if ( ping_receiver == 0 || ping_sender == 0 ) {
      return -1; 
    }
  }
  else {
    if ( ping_receiver == 0 ) {
      return -1; 
    }
  }

  if ( rc_snid == 0 || rc_rnid == 0 ) {
    return -1; 
  }

  return 0;
}


void
open_dev(int sd0, int sd1, char buf[], int bufsize, char* str)
{
  int len;

  /* try to open /dev/portals */
  write( sd0, "y", 1 );            /* proceed */
  len = read(sd0, buf, bufsize);
  buf[len] = 0;

  if ( !strncmp("n", buf, strlen("n")) ) {
    printf("vroute: %s failed to open /dev/portals\n", str);
    write( sd1, "n", 1 );          /* kill receiver */
    exit(-1);
  printf("------------------------------------------\n");
  }
  printf("vroute: %s opened /dev/portals\n", str);

  /* try to open /dev/rtscts */
  write( sd0, "y", 1 );            /* proceed */
  len = read(sd0, buf, bufsize);
  buf[len] = 0;

  if ( !strncmp("n", buf, strlen("n")) ) {
    printf("vroute: %s failed to open /dev/rtscts\n", str);
    write( sd1, "n", 1 );          /* kill receiver */
    printf("------------------------------------------\n");
    exit(-1);
  }
  printf("vroute: %s opened /dev/rtscts\n", str);
}


int
nid_check(int sd0, char buf[], int bufsize, int my_nid, int* nid, 
          char* str, int verify)
{
  int len, rc=1;

  write( sd0, "y", 1 );            /* proceed */
//len = read(sd0, buf, bufsize);
  len = doRead(sd0, buf, bufsize, sizeof(istr));
  buf[len] = 0;

 *nid = strtol(buf, NULL, 10);
//printf("vroute: %s: nid, my_nid= %d, %d\n", str, *nid, my_nid);

  if ( verify ) {
    if ( *nid != my_nid ) {
      rc = 0;
      printf("vroute: %s and i disagree on his nid: i  say  %d\n", str, my_nid);
      printf("vroute: %s and i disagree on his nid: he says %d\n", str, *nid);
    }
    else {
      printf("vroute: %s and i agree on his nid %d: verified.\n", str, *nid);
    }
  }
  return rc;
}


int do_ping(int send_sd, int recv_sd, char send_buf[], char recv_buf[], 
            int bufsize, int snid __attribute__ ((unused)), int rnid, 
            char* send_str, char* recv_str)
{
  int rlen, slen, ping_ok;

  /* tell receiver to proceed -- clear pINgFO[snid] */

  write( recv_sd, "y", 1 );            /* proceed */
  rlen = read(recv_sd, recv_buf, bufsize);
  printf("vroute: %s cleared ping_info[%s]\n", recv_str, send_str);

  /* tell sender to proceed -- ping receiver */
 
  write( send_sd, "y", 1 );            /* proceed */
  slen = read(send_sd, send_buf, bufsize);
  printf("vroute: %s tried to ping %s --\n", send_str, recv_str); 
  printf("vroute: wait a bit before telling %s to check for incoming msg...\n", recv_str);

  /* sleep for a couple of seconds -- else have receiver poll
     pINgFO[snid] and timeout...
  */
  sleep(no_sleep_secs);

  /* tell receiver to proceed -- check pINgFO[snid]  for change */
  write( recv_sd, "y", 1 ); 
  rlen = read(recv_sd, recv_buf, bufsize);

  if ( !strncmp("y", recv_buf, strlen("y")) ) {
    printf("vroute: PING SUCCEEDED: %s got ping from %s...\n", recv_str, send_str);
    ping_ok = 1;
  }
  else {
    printf("vroute: PING FAILED: %s did not see change in ping_info[%s]\n", recv_str, send_str);
    ping_ok = 0;
  }
  return ping_ok;
}



int self_ping(int send_sd, char send_buf[], int bufsize,
              int snid __attribute__ ((unused)), char* send_str)
{
  int slen, ping_ok;

  /* tell sender to proceed -- clear pINgFO[snid] */

  write( send_sd, "y", 1 );            /* proceed */
  slen = read(send_sd, send_buf, bufsize);
  printf("vroute: %s cleared ping_info[%s]\n", send_str, send_str);

  /* tell sender to proceed -- ping self */
 
  write( send_sd, "y", 1 );            /* proceed */
  slen = read(send_sd, send_buf, bufsize);
  printf("vroute: %s tried to ping self --\n", send_str); 
  printf("vroute: wait a bit before telling %s to check for incoming msg...\n", send_str);

  /* sleep for a couple of seconds -- else have receiver poll
     pINgFO[snid] and timeout...
  */
  sleep(2);

  /* tell sender to proceed -- check pINgFO[snid]  for change */
  write( send_sd, "y", 1 ); 
  slen = read(send_sd, send_buf, bufsize);

  if ( !strncmp("n", send_buf, strlen("n")) ) {
    printf("vroute: PING FAILED: %s did not see change in ping_info[%s]\n", send_str, send_str);
    ping_ok = 0;
  }
  else {
    printf("vroute: PING SUCCEEDED: %s got ping from %s...\n", send_str, send_str);
    ping_ok = 1;
  }
  return ping_ok;
}


int check_args(int got_sname, int got_rname)
{
  int fail = 0;

  if ( !got_sname ) {
    printf("vroute: ERROR -- missing cmd line arg for sender host name!\n");
    fail = -1;
  }

  if ( !got_rname ) {
    printf("vroute: ERROR -- missing cmd line arg for receiver host name!\n");
    fail = -1;
  }

  if (fail == -1) {
    show_usage();
  }

  return fail;
}

void show_usage()
{
  printf("usage:\n");
  printf("vroute runs on an SSS node...\n");
  printf("  -f, -from  nodename   -- do myrinet ping from this node (required)\n");
  printf("  -t, -to    nodename   -- do myrinet ping to   this node (required)\n");
  printf("  -b, -both             -- attempt ping in both directions (optional)\n");
  printf("  -p, -port             -- specify vrouted port number (optional)\n");

  printf("  -s, -snid  node id    -- verify the sender's   node id (optional)\n");
  printf("  -r, -rnid  node id    -- verify the receiver's node id (optional)\n");
  printf("  -z,        no_secs    -- number of seconds we sleep between asking\n");
  printf("                        -- sender to send and receiver to check for receipt.\n");
  printf("                        -- the default is 1 second (optional)\n");

  printf("  -v, -verbosity        -- get verbose output (optional)\n");

  printf("  -h, -help             -- this stuff (optional)\n");

}
